opm/grid/common/SubGridView.hpp

```diff
@@ -0,0 +1,371 @@
+                return index_ != other.index_;
+            }
+        private:
+            const SubGridView* view_;
+            std::size_t index_;
+            mutable Entity entity_; // This may be low-performing for grids with large Entity objects.
+        };
+        using Iterator = SubIterator;
+
+        /** \brief Define types needed to iterate over entities of a given partition type */
+        template <PartitionIteratorType pit>
+        struct Partition {
+            /** \brief iterator over a given codim and partition type */
+            using Iterator = SubIterator;
+        };
+
+    };
+
+    enum { conforming = Traits::conforming };
+    enum { dimension = GridImp::dimension };
+
+public:
+
+    /// Construct a view of the codim 0 entities that can be constructed from the seeds input.
+    ///
+    /// The seeds input is moved from and will be in a valid but indeterminate state after the call.
+    SubGridView(const Grid& grid,
+                std::vector<typename Codim<0>::Entity::EntitySeed>&& seeds,
+                const bool overlap = true)
+        : grid_(&grid)
+        , subset_(std::move(seeds))
+        , num_owned_(subset_.size())
+    {
+        // Nothing more to do if we do not want to have overlap entities.
+        if (!overlap) {
+            return;
+        }
+
+        // Add neighbouring not-owned entities to subset_
+        using Seed = typename Codim<0>::Entity::EntitySeed;
+        std::unordered_set<int> owned;
+        std::unordered_map<int, Seed> neighbors;
+        const auto& iset = grid_->leafIndexSet();
+        const auto& leaf_view = grid_->leafGridView();
+        for (const auto& seed : subset_) {
+            // Add this entity to the set of owned indices.
+            const auto& entity = grid_->entity(seed);
+            owned.insert(iset.index(entity));
+            // Iterating over all intersections, ...
+            const auto end = leaf_view.iend(entity);
+            for (auto it = leaf_view.ibegin(entity); it != end; ++it) {
+                if (it->boundary()) {
+                    continue;
+                }
+                if (it->neighbor()) {
+                    const auto outside_entity = it->outside();
+                    // ...for all neighbour entities, add to neighbors.
+                    neighbors.try_emplace(iset.index(outside_entity), outside_entity.seed());
+                }
+            }
+        }
+        // Now that owned is complete, we can eliminate any owned entries.
+        std::map<int, Seed> unowned_neighbors;
+        for (const auto& nb : neighbors) {
+            if (owned.count(nb.first) == 0) {
+                unowned_neighbors.insert(nb);
+            }
+        }
+        subset_.resize(subset_.size() + unowned_neighbors.size());
+        std::size_t count = num_owned_;
+        for (const auto& [index, seed] : unowned_neighbors) {
+            subset_[count] = seed;
+            ++count;
+        }
```

Comment:
Come to think of if, this section could be reduced to
```C++
for (const auto& [index, seed] : neighbors) {
    if (owned.count(index) == 0) {
        this->subset_.push_back(seed);
    }
}
this->subset_.shrink_to_fit();
```

Reply:
That would give a random order to the neighbours, I would like them to have the predictable order (same as global index order). It could also be important to maintain the "owned by process first" order within the subset overlap partition when we go to parallel, although that is not really thought through.

This is also why the unowned_neighbors is a map and not unordered_map.